pulsar-tracing/src/main/java/com/datastax/oss/pulsar/tracing/BrokerTracing.java

```diff
@@ -0,0 +1,645 @@
+        log.debug("Unknown tracing level: {}", level);
+        break;
+    }
+  }
+
+  private static void populatePublishContext(
+      Topic.PublishContext publishContext, Map<String, Object> traceDetails) {
+    traceDetails.put("publishContext.isMarkerMessage", publishContext.isMarkerMessage());
+    traceDetails.put("publishContext.isChunked", publishContext.isChunked());
+    traceDetails.put("publishContext.numberOfMessages", publishContext.getNumberOfMessages());
+
+    traceDetails.put("publishContext.entryTimestamp", publishContext.getEntryTimestamp());
+    traceDetails.put("publishContext.msgSize", publishContext.getMsgSize());
+    traceDetails.put("publishContext.producerName", publishContext.getProducerName());
+    traceDetails.put(
+        "publishContext.originalProducerName", publishContext.getOriginalProducerName());
+    traceDetails.put("publishContext.originalSequenceId", publishContext.getOriginalSequenceId());
+    traceDetails.put("publishContext.sequenceId", publishContext.getSequenceId());
+  }
+
+  private static void traceByteBuf(String key, ByteBuf buf, Map<String, Object> traceDetails) {
+    if (buf == null) return;
+
+    if (buf.readableBytes() < MAX_DATA_LENGTH) {
+      traceDetails.put(key, Hex.encodeHex(buf.nioBuffer()));
+    } else {
+      traceDetails.put(key + "Slice", Hex.encodeHex(buf.slice(0, MAX_DATA_LENGTH).nioBuffer()));
+    }
+  }
+
+  private static TraceLevel getTracingLevel(Subscription sub) {
+    if (sub == null
+        || sub.getSubscriptionProperties() == null
+        || !sub.getSubscriptionProperties().containsKey("trace")) {
+      return TraceLevel.NONE;
+    }
+    try {
+      return TraceLevel.valueOf(sub.getSubscriptionProperties().get("trace"));
+    } catch (IllegalArgumentException e) {
+      log.debug(
+          "Invalid tracing level: {}. Setting to NONE for subscription {}",
+          sub.getSubscriptionProperties().get("trace"),
+          sub);
+      return TraceLevel.NONE;
+    }
+  }
+
+  private static TraceLevel getTracingLevel(Producer producer) {
+    if (producer == null
+        || producer.getMetadata() == null
+        || !producer.getMetadata().containsKey("trace")) {
+      return TraceLevel.NONE;
+    }
+    try {
+      return TraceLevel.valueOf(producer.getMetadata().get("trace"));
+    } catch (IllegalArgumentException e) {
+      log.debug(
+          "Invalid tracing level: {}. Setting to NONE for producer {}",
+          producer.getMetadata().get("trace"),
+          producer);
+      return TraceLevel.NONE;
+    }
+  }
+
+  //    private boolean needToTraceTopic(Topic topic) {
+  //        if (topic == null) return false;
+  //
+  //        // todo: how to read topic props
+  //        return true;
+  //    }
+
+  /* ***************************
+   **  Administrative events
+   ******************************/
+
+  public void onConnectionCreated(ServerCnx cnx) {
+    if (!getEnabledEvents(cnx).contains(EventReasons.ADMINISTRATIVE)) return;
```

Comment:
if this is something configured only once, we can have a boolean flag "traceConnections", and we will save CPU cycles